move is_field from Ring to Rings

mantepse opened the pull request:
We move the method `is_field` from the class `Ring` to the category `Rings`

mantepse committed:
move is_field from Ring to Rings

mantepse committed:
adapt doctest

mantepse committed:
assert that a coercion cannot be registered after a conversion has been discovered, disable check for zero ring in is_field

mantepse committed:
take category into account when checking whether codomain is valid in is_NumberFieldHomsetCodomain

mantepse committed:
Merge branch 'sagemath:develop' into move_is_field

mantepse committed:
revert commenting out code in is_field

mantepse committed:
Merge branch 'move_is_field' of https://github.com/mantepse/sage into move_is_field

saraedum committed:
Allow number field morphisms to rings that are not fields

saraedum commented on src/sage/categories/magmas.py:
```diff
@@ -379,21 +379,18 @@ def is_field(self, proof=True):
-                if not self.base_ring().is_field(proof):
-                    return False
-                return (self.basis().keys().cardinality() == 1)
+                raise NotImplementedError
```
Shouldn't this be an abstract method instead? I think that every ring must implement this (even if it just returns `NotImplementedError` in some rare cases where we cannot decide.

tscrim replied:
It would be a `@abstract_method(optional=True)` method as currently implemented. I don’t think it should be required as it might not be computationally feasible (e.g., equivalent to solving the word problem).

Actually, for a magma with 1 element, it must be a monoid since the multiplication must be an element back in the magma with $x^2 = x$ necessarily. So the second part of the docstring is unnecessary.

fchapoton replied:
I think that no ring should implement "is_field". Every ring should rather be put in the proper category at the init time.

saraedum replied:
@fchapoton, I believe that the idea here is that for some rings it's very expensive to decide this, e.g., Z/nZ with a very big n. That's why you don't want to determine the category at init time I think.

saraedum replied:
> It would be a `@abstract_method(optional=True)` method as currently implemented. I don’t think it should be required as it might not be computationally feasible (e.g., equivalent to solving the word problem).

I propose that it returns `NotImplementedError` in such cases.

tscrim replied:
This is what the `optional=True` is meant for.

saraedum replied:
Yes, both work and lead to the same result. My understanding was that `optional=True` is for things that make sense but might be tricky to implement, say `quo_rem` or factorization. If you essentially want to force all parents to implement it (and only allow isolated cases to opt-out) you use `optional=False`.
It's really just a detail, and I think we're going off-topic a bit here but the main difference here is in testing code. In one case the `_test_is_field` does `if is_field == NotImplemented` in the other case you'd probably have to explicitly disable `_test_is_field` for your parent.
No strong feelings here from my side. I just used that convention in the past, but it's quite possible that that's just my understanding of how things work.

saraedum commented on src/sage/categories/rings.py:
```diff
@@ -556,6 +556,54 @@ def is_subring(self, other):
             except (TypeError, AttributeError):
                 return False
 
+        def is_field(self, proof=True):
```
Have we tried to see what happens if we just delete this? (Maybe it makes the world a better place…)

saraedum committed:
Set correct category for morphisms from number fields

saraedum committed:
Adapt is_NumberFieldHomsetCodomain doctests to recent changes in implementation

saraedum committed:
Partially revert 10c5abf85b9c0e18b2bb635d2a5f9ed7f3fc0096

on my system, even before #39520, this ring is reported as commutative
rings and not as fields. Not sure when this changed but it seems that it
makes sense to revert this to what the book says.

saraedum committed:
Fix lint

saraedum commented on src/sage/categories/monoids.py:
```diff
@@ -504,6 +504,26 @@ def extra_super_categories(self):
+                if not self.base_ring().is_field(proof):
+                    return False
+                return (self.basis().keys().cardinality() == 1)
+
```
Why was this moved to monoids? It seems that docstrings between this one and the magma one are out of sync.

mantepse replied:
If anybody knows, then it is @fchapoton.  I think that we tried to make `SymmetricGroup(1).algebra(QQ).is_field()` work.

That said, I also think that `is_field` should be an abstract method, and not try to detect the one trivial case where it's return value can be determined.

tscrim replied:
We can remove this method since the one from the magmatic algebras is necessarily exactly the same check.

fchapoton replied:
The idea is that (one day) the only place defining "is_field" will be in the categories. This seems to be inconsistent with the idea of abstract methods.

saraedum committed:
Merge branch 'develop' into move_is_field

saraedum commented on src/sage/categories/monoids.py:
```diff
@@ -504,6 +504,26 @@ def extra_super_categories(self):
+                if not self.base_ring().is_field(proof):
+                    return False
+                return (self.basis().keys().cardinality() == 1)
+
```
```suggestion
```

saraedum commented on src/sage/categories/magmas.py:
```diff
@@ -379,21 +379,18 @@ def is_field(self, proof=True):
+                    sage: Magmas().example().algebra(QQ).is_field()
+                    Traceback (most recent call last):
+                    ...
+                    NotImplementedError
```
```suggestion
                    sage: SymmetricGroup(1).algebra(QQ).is_field()                      # needs sage.combinat sage.groups
                    True
                    sage: SymmetricGroup(1).algebra(ZZ).is_field()                      # needs sage.combinat sage.groups
                    False
                    sage: SymmetricGroup(2).algebra(QQ).is_field()                      # needs sage.combinat sage.groups
                    False
                    sage: Magmas().example().algebra(QQ).is_field()
                    Traceback (most recent call last):
                    ...
                    NotImplementedError
```

saraedum committed:
Merge is_field of magmas and monoids

since monoids are magmas there seems to be no point in keeping the separation.

saraedum committed:
Restore is_field() implementation for magma algebras

saraedum committed:
Improve documentation of magma algebras

if there's only one element in the magma, it's square cannot be "zero" I
think.

saraedum committed:
Fix magma algebra doctest

tscrim commented on src/sage/categories/magmas.py:
```diff
@@ -390,10 +391,13 @@ def is_field(self, proof=True):
                     False
                     sage: SymmetricGroup(2).algebra(QQ).is_field()                      # needs sage.combinat sage.groups
                     False
+                    sage: Magmas().example().algebra(QQ).is_field()
```
This test is easy to fix by adding `.Infinite()` to the category in the `__init__` of the example.

tscrim commented on src/sage/rings/number_field/number_field.py:
```diff
@@ -156,24 +152,32 @@ def is_NumberFieldHomsetCodomain(codomain):
+        return False
+
+    if codomain not in category:
+        return False
```
```suggestion
    assert codomain in category
```
I believe we agreed on this.

mantepse replied:
I tried this, but then I get
```
sage -t --warn-long 5.0 --random-seed=119788034856104425102958228681385469939 src/sage/arith/misc.py
**********************************************************************
File "src/sage/arith/misc.py", line 2060, in sage.arith.misc.xgcd
Failed example:
    xgcd(y^2, a*h*y + b)
Exception raised:
    Traceback (most recent call last):
      File "sage/misc/cachefunc.pyx", line 1970, in sage.misc.cachefunc.CachedMethodCaller.__call__
        return cache[k]
    KeyError: (Number Field in g with defining polynomial x^2 - 3, None, True)

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/martin/sage/src/sage/doctest/forker.py", line 728, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/martin/sage/src/sage/doctest/forker.py", line 1152, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.arith.misc.xgcd[34]>", line 1, in <module>
        xgcd(y**Integer(2), a*h*y + b)
      File "/home/martin/sage/src/sage/arith/misc.py", line 2079, in xgcd
        return a.xgcd(b)
               ^^^^^^^^^
      File "sage/structure/element.pyx", line 4743, in sage.structure.element.coerce_binop.new_method
        return method(self, other, *args, **kwargs)
      File "sage/rings/polynomial/polynomial_element.pyx", line 9709, in sage.rings.polynomial.polynomial_element.Polynomial.xgcd
        return self.base_ring()._xgcd_univariate_polynomial(self, other)
      File "/home/martin/sage/src/sage/categories/fields.py", line 466, in _xgcd_univariate_polynomial
        q, r = d.quo_rem(v3)
               ^^^^^^^^^^^^^
      File "sage/structure/element.pyx", line 4743, in sage.structure.element.coerce_binop.new_method
        return method(self, other, *args, **kwargs)
      File "sage/rings/polynomial/polynomial_element.pyx", line 12538, in sage.rings.polynomial.polynomial_element.Polynomial_generic_dense.quo_rem
        q = x[n+k-1] * inv
      File "sage/structure/element.pyx", line 1505, in sage.structure.element.Element.__mul__
        return (<Element>left)._mul_(right)
      File "sage/rings/fraction_field_element.pyx", line 705, in sage.rings.fraction_field_element.FractionFieldElement._mul_
        d1 = rnum.gcd(sden)
      File "sage/rings/polynomial/multi_polynomial.pyx", line 2222, in sage.rings.polynomial.multi_polynomial.MPolynomial.gcd
        return self.parent()(doit(uniself, other.polynomial(x)))
      File "/home/martin/sage/src/sage/categories/unique_factorization_domains.py", line 220, in _gcd_univariate_polynomial
        _, R = A.pseudo_quo_rem(B)
               ^^^^^^^^^^^^^^^^^^^
      File "sage/rings/polynomial/polynomial_element.pyx", line 5409, in sage.rings.polynomial.polynomial_element.Polynomial.pseudo_quo_rem
        if other in self._parent.base_ring():
      File "sage/structure/parent.pyx", line 1170, in sage.structure.parent.Parent.__contains__
        x2 = self(x)
      File "sage/structure/parent.pyx", line 901, in sage.structure.parent.Parent.__call__
        return mor._call_(x)
      File "sage/rings/polynomial/polynomial_element.pyx", line 12926, in sage.rings.polynomial.polynomial_element.ConstantPolynomialSection._call_
        raise TypeError(f"{x} is not a constant polynomial")
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3190, in sage.rings.polynomial.polynomial_element.Polynomial._repr_
        return self._repr()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3154, in sage.rings.polynomial.polynomial_element.Polynomial._repr
        x = y = repr(x)
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3190, in sage.rings.polynomial.polynomial_element.Polynomial._repr_
        return self._repr()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3154, in sage.rings.polynomial.polynomial_element.Polynomial._repr
        x = y = repr(x)
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/number_field/number_field_element.pyx", line 5052, in sage.rings.number_field.number_field_element.NumberFieldElement_relative._repr_
        return repr(R(self.list()))
      File "sage/rings/number_field/number_field_element.pyx", line 5019, in sage.rings.number_field.number_field_element.NumberFieldElement_relative.list
        return self.vector().list()
      File "sage/rings/number_field/number_field_element.pyx", line 3675, in sage.rings.number_field.number_field_element.NumberFieldElement.vector
        return self.number_field().relative_vector_space()[2](self)
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 1456, in relative_vector_space
        return self.free_module(self.base_field(), *args, **kwds)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/misc/cachefunc.pyx", line 1975, in sage.misc.cachefunc.CachedMethodCaller.__call__
        w = self._instance_call(*args, **kwds)
      File "sage/misc/cachefunc.pyx", line 1851, in sage.misc.cachefunc.CachedMethodCaller._instance_call
        return self.f(self._instance, *args, **kwds)
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 1413, in free_module
        to = maps.MapRelativeNumberFieldToRelativeVectorSpace(self, V)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/maps.py", line 346, in __init__
        NumberFieldIsomorphism.__init__(self, Hom(K, V))
                                              ^^^^^^^^^
      File "/home/martin/sage/src/sage/categories/homset.py", line 418, in Hom
        H = Hom(X, Y, category, check=False)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/categories/homset.py", line 445, in Hom
        H = X._Hom_(Y, category)
            ^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 778, in _Hom_
        if not is_NumberFieldHomsetCodomain(codomain):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/number_field.py", line 177, in is_NumberFieldHomsetCodomain
        assert codomain in category, f"codomain (={codomain}) should be in the category (={category})"
               ^^^^^^^^^^^^^^^^^^^^
    AssertionError: codomain (=Vector space of dimension 2 over Number Field in g with defining polynomial x^2 - 3) should be in the category (=Category of rings)
```

mantepse replied:
(and many similar failures in `schemes`)

tscrim replied:
I think we need to be more careful about the default category. Rings is a bad choice as it doesn't take into account the inputs. I think it should be the meet of the (co)domain's categories.

fchapoton replied:
@tscrim : could you please turn your wish into a more precise suggestion ?

tscrim replied:
I added it above (since only the codomain is given to the method here).

tscrim replied:
My change above is also with making this an assert statement.

tscrim replied:
I still would like this changed to an `assert` statement:
```suggestion
    assert codomain in category
```

tscrim reviewed (changes requested):
The symmetric group tests are failing because the `is_field` in `Rings()` is taking priority (i.e., it is higher in the MRO):
```
 ...
 Category of monoid algebras over Rational Field,
 ...
 Category of rings,
 Category of semigroup algebras over Rational Field,
 ...
 Category of magmatic algebras over Rational Field,
 ...
```
However, this is the correct ordering. Ah, now I know why @fchapoton had the implementation in `Monoids.Algebras`, that is a subcategory of `Rings` and needs to override it. There probably is an easier way to do things than copy the full method...

mantepse committed:
add the implementation of is_field for monoid algebras, so that it takes priority over the implementation in Rings

mantepse committed:
a FreeMagma is infinite

fchapoton commented on src/sage/categories/examples/magmas.py:
```diff
@@ -73,7 +72,7 @@ def __init__(self, alphabet=('a', 'b', 'c', 'd')):
                              "'(', ')' or '*'")
         self.alphabet = alphabet
-        Parent.__init__(self, category=Magmas().FinitelyGenerated())
+        Parent.__init__(self, category=Magmas().FinitelyGenerated().Infinite())
```
well, unless the set of generators is empty..

mantepse replied:
I'll just wait until the bots are done, then I'd apply
```suggestion
        cat = Magmas().FinitelyGenerated()
        if alphabet:
            Parent.__init__(self, category=cat.Infinite())
        else:
            Parent.__init__(self, category=cat)
```

mantepse replied:
Actually, it might be better to require that there is at least one generator.  Otherwise, it would be necessary to somehow state that the magma is empty.

saraedum committed:
Require free magmas to have at least one generator

so that the category (and other logic) is correct

saraedum committed:
Merge branch 'develop' into move_is_field

saraedum committed:
Merge branch 'develop' into move_is_field

fchapoton committed:
Update monoids.py

tscrim commented on src/sage/rings/number_field/number_field.py:
```diff
@@ -156,24 +152,32 @@ def is_NumberFieldHomsetCodomain(codomain):
+    from sage.categories.rings import Rings
+
+    if category is None:
+        category = Rings()
```
```suggestion
        category = codomain.category()
```

fchapoton replied:
ok, thanks,  looks good, let's try

fchapoton committed:
Update src/sage/rings/number_field/number_field.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

fchapoton committed:
Merge branch 'develop' into move_is_field

fchapoton committed:
Merge branch 'develop' into move_is_field

fchapoton committed:
change to assert

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

fchapoton committed:
Merge branch 'develop' into move_is_field

fchapoton reviewed (approved):
let's move on
